File: backend/src/OpenQuiz.Infrastructure/Services/ScoreService.cs
using Microsoft.EntityFrameworkCore;
using OpenQuiz.Application.Abstractions;
using OpenQuiz.Application.Scores;
using OpenQuiz.Infrastructure.Persistence;

namespace OpenQuiz.Infrastructure.Services;

public class ScoreService : IScoreService
{
    private readonly OpenQuizDbContext _db;
    public ScoreService(OpenQuizDbContext db) => _db = db;

    public async Task<List<ScoreEntry>> LeaderboardAsync(Guid? pollId, int top, CancellationToken ct)
    {
        var q = _db.Scores.AsNoTracking();
        q = pollId.HasValue ? q.Where(s => s.PollId == pollId) : q.Where(s => s.PollId == null);

        var list = await q.OrderByDescending(s => s.Points).ThenBy(s => s.TotalTimeMs)
            .Take(Math.Clamp(top, 1, 500))
            .Select(s => new ScoreEntry(s.UserName, s.Points, s.TotalTimeMs))
            .ToListAsync(ct);
        return list;
    }
}

File: frontend/src/services/scoreService.js
import { api } from './apiClient';

export const scoreService = {
  leaderboard: (pollId, top = 100) => {
    const qs = new URLSearchParams();
    if (pollId) qs.set('pollId', pollId);
    qs.set('top', top);
    return api.get(`/api/scores?${qs.toString()}`, { auth: false });
  },
};
